subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectListTest.groovy

```diff
@@ -396,6 +406,31 @@ class DefaultNamedDomainObjectListTest extends AbstractNamedDomainObjectCollecti
         list.findAll { it != "b" } == ["a", "c"]
     }
 
+    def "name based filtering does not realize pending"() {
+        given:
+        list.add("realized1")
+        list.addLater(new NamedStringProvider("unrealized1"))
+        list.add("realized2")
+        list.addLater(new NamedStringProvider("unrealized2"))
+        println "something"
+
+        expect: "unrealized elements remain as such"
+        list.index.asMap().size() == 2
+        list.index.pendingAsMap.size() == 2
+
+        when:
+        def filtered = list.named { it.contains("2") }
+
+        then: "unrealized elements remain as such"
+        list.index.asMap().size() == 2
+        list.index.pendingAsMap.size() == 2
+
+        filtered.index.asMap().size() == 1
+        filtered.index.pendingAsMap.size() == 1
+
+        filtered == ["realized2", "unrealized2"]
```

Comment:
Need to test that iterating over `filtered` only realizes the elements actually in filtered, not other, excluded ones too from the original.